### PR TITLE
HPCC-28019 Add esp service to get the hotspots for a workunit

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@GordonSmith Does this look like what you would hope for?

**Comment by ghalliday** [comment]
Provisional version.  Many of the other analysis tools need revising before publishing as services (see HPCC-28020, HPCC-28022)

File: esp/scm/ws_workunits_req_resp.ecm [hunk 1]
```diff
@@ -1092,4 +1096,37 @@ ESPresponse [exceptions_inline] WUGetPluginsResponse
 {
     ESParray<ESPstruct WUEclPluginsInFolder> Plugins;
 };
+
+// ----------------------------------------------------------------------------------
+
+ESPrequest [nil_remove] WUAnalyseHotspotRequest
+{
+    string Wuid;
+    string RootScope;
```
Comment (GordonSmith) [inline esp/scm/ws_workunits_req_resp.ecm:1105]: Is this optional? 
Should it default to the very root?

File: esp/scm/ws_workunits_req_resp.ecm [hunk 2]
```diff
@@ -1092,4 +1096,37 @@ ESPresponse [exceptions_inline] WUGetPluginsResponse
...
+
+ESPStruct WUHotspotItem
+{
+    string Scope;
+    string ParentScope;
+    double TotalTimePercent;
+    double StartTimePercent;
+    double StartAndDependentsTimePercent;
+    double ExecuteTimePercent;
+};
```
Comment (GordonSmith) [inline esp/scm/ws_workunits_req_resp.ecm:1122]: This doesn't appear to be returned in any response?

File: esp/scm/ws_workunits_req_resp.ecm (hunk 1 above)
Comment (ghalliday) [reply esp/scm/ws_workunits_req_resp.ecm:1105]: yes it should be optional

File: esp/scm/ws_workunits_req_resp.ecm (hunk 2 above)
Comment (ghalliday) [reply esp/scm/ws_workunits_req_resp.ecm:1122]: I think it can be deleted.  This was before our discussion, and results now returned in wudetails format.

**Comment by ghalliday** [comment]
@shamser this is the PR I mentioned.  I think it is ready for a first review.

**Review by shamser — CHANGES_REQUESTED** [review]

File: common/wuanalysis/anawu.cpp [hunk 3]
```diff
@@ -241,7 +261,100 @@ bool WuScopeHashTable::matchesElement(const void *et, const void *searchET) cons
...
+}
+
+void WuHotspotResult::reportTime()
+{
+    const char * prefix = isRoot ? "R" : " ";
+    ThorActivityKind kind = activity->queryThorActivityKind();
+    const char * critical = activity->  isOnCriticalPath() ? "*" : " ";
+    printf("%s%-6s {%.9f..%.9f,%.9f..%.9f} %s(%-6s %6.2f%%,%6.2f%%,%6.2f%%) %s\n",
+            critical, activity->queryName(),
+            activity->getBeginTimestamp() / 1E9, activity->getEndStartTimestamp() / 1E9, activity->getBeginRunTimestamp() / 1E9, activity->getEndTimestamp() / 1E9,
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp:345]: 1E9 is a double c++ constant - so the divide will be done as a double.

File: common/wuanalysis/anawu.cpp [hunk 4]
```diff
@@ -521,3 +2083,212 @@ void WUANALYSIS_API analyseAndPrintIssues(IConstWorkUnit * wu, double costRate,
...
+    analyser.findHotspots(cfg->queryProp("@rootScope"), results.totalTime, results.hotspots);
+    results.root.setown(analyser.getRootScope());
+}
+
+/*
+Useful things to know:
+
+Using wutool to investigate query performance.
+
+- If using a woorkunit that was generated to store the stats, then use that workunit. E.g.
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp]: fixed

File: common/wuanalysis/anawu.cpp [hunk 5]
```diff
@@ -21,81 +21,45 @@
...
+// The following are set from options in the query
+    bool onlyActive = true;
+    double thresholdPercent = 1.0;
+    bool onlyCriticalPath = false;
+
+//The following are computed:
+    stat_type minTime = 0;
+    stat_type maxTime = (stat_type)-1;
+    stat_type timeThreshold = 0;
+    PointerArrayOf<WaThread> threadFilter;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp]: added a comment.  Not actually used at the moment.

File: common/wuanalysis/anawu.cpp [hunk 6]
```diff
@@ -110,11 +74,13 @@ constexpr struct WuOption wuOptionsDefaults[watOptMax]
     {watOptMinInterestingCost, "minInterestingCost", 30000, wutOptValueTypeMSec},
     {watOptSkewThreshold, "skewThreshold", 20, wutOptValueTypePercent},
     {watOptMinRowsPerNode, "minRowsPerNode", 1000, wutOptValueTypeCount},
-    {watPreFilteredKJThreshold, "preFilteredKJThreshold", 50, wutOptValueTypePercent} };
+    {watPreFilteredKJThreshold, "preFilteredKJThreshold", 50, wutOptValueTypePercent},
+    {watIsRoxie, "isRoxie", true, wutOptValueTypeBool},
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp]: deleted - it was unused.

File: common/wuanalysis/anawu.cpp [hunk 7]
```diff
@@ -262,257 +375,1706 @@ void WuScope::connectActivities()
...
-    }
-    virtual void noteAttribute(WuAttr attr, const char * value)
+    for (auto input : inputs)
     {
-        StringBuffer name;
-        name.append('@').append(queryWuAttributeName(attr));
-        scope->setProp(name, value);
+        WuScope * source = input->querySource();
+        stat_type childTime = source->getStartTime();
+        if (childTime * 10 > startupTime)
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp:611]: Added a comment.  This code isn't currently used by the main service.

File: common/wuanalysis/anawu.cpp [hunk 8]
```diff
@@ -262,257 +375,1706 @@ void WuScope::connectActivities()
...
+        }
+    }
+    else
+    {
+        //Exclude inputs that are not on the critical path.  Either they are started after the first input finishes,
+        //or they are started after the child dependency we are tracking finishes.  (The dependency will always be lower)
+        stat_type maxTime = maxDependantTime;
+        if ((maxTime == 0) && (callers.size() > 1))
+        {
+        }
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/wuanalysis/anawu.cpp]: deleted

**Review by shamser — APPROVED** [review]

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**